speed_of_light.go

```diff
@@ -0,0 +1,97 @@
+// 2024, michae2
+
+package main
+
+import (
+	"fmt"
+	"math"
+	"os"
+	"runtime"
+	"runtime/debug"
+	"sync"
+	"sync/atomic"
+	"syscall"
+)
+
+// how fast can we do just the io part?
+
+const chunk = 1 << 26
+
+func main() {
+	//syscall.Setpriority(syscall.PRIO_PROCESS, 0, -20)
+
+	// disable GC
+	debug.SetMemoryLimit(math.MaxInt64)
+	debug.SetGCPercent(-1)
+
+	/*
+		pfile, err := os.Create("cpu.prof")
+		if err != nil {
+			panic(err)
+		}
+		defer pfile.Close()
+		pprof.StartCPUProfile(pfile)
+		defer pprof.StopCPUProfile()
+	*/
+
+	// mmap the measurements file
+	file, err := os.Open("measurements.txt")
+	if err != nil {
+		panic(err)
+	}
+	defer file.Close()
+	fd := int(file.Fd())
+
+	fileInfo, err := file.Stat()
+	if err != nil {
+		panic(err)
+	}
+	fileSize := int(fileInfo.Size())
+
+	data, err := syscall.Mmap(fd, 0, fileSize, syscall.PROT_READ, syscall.MAP_SHARED)
+	if err != nil {
+		panic(err)
+	}
+	defer syscall.Munmap(data)
+
+	workers := runtime.NumCPU()
+
+	// build a queue of file chunks
+	chunks := make(chan []byte, len(data)/chunk+1+workers)
+	for i := 0; i < len(data); i += chunk {
+		next := i + chunk
+		if next > len(data) {
+			next = len(data)
+		}
+		chunks <- data[i:next]
+	}
+	for i := 0; i < workers; i++ {
+		chunks <- nil
+	}
+
+	// run workers
+	var total uint64
+	var wait sync.WaitGroup
+	for i := 0; i < workers; i++ {
+		wait.Add(1)
+		go parse(chunks, &total, &wait)
+	}
+
+	wait.Wait()
+	fmt.Println(total)
+}
+
+func parse(chunks <-chan []byte, total *uint64, wait *sync.WaitGroup) {
+	defer wait.Done()
+	var subtotal uint64
+	for subdata := range chunks {
+		if subdata == nil {
+			// no more chunks
+			atomic.AddUint64(total, subtotal)
+			return
+		}
+		for _, b := range subdata {
+			subtotal += uint64(b)
```

Comment:
Not quite speed of light IMO because you're only processing 1 byte at a time. I think a better speed of light test is counting the number of newlines in the file using `bytes.Count` which has an assembly optimized implementation.

Reply:
Great point!